File: cpp/Dynamic Programming/neighborhood_burglary_optimized.cpp
#include <vector>
#include <algorithm>

int neighborhoodBurglaryOptimized(std::vector<int>& houses) {
    if (houses.empty()) {
        return 0;
    }
    if (houses.size() == 1) {
        return houses[0];
    }
    // Initialize the variables with the base cases.
    int prevMaxProfit = std::max(houses[0], houses[1]);
    int prevPrevMaxProfit = houses[0];
    for (int i = 2; i < houses.size(); i++) {
        int currMaxProfit = std::max(prevMaxProfit, houses[i] + prevPrevMaxProfit);
        // Update the values for the next iteration.
        prevPrevMaxProfit = prevMaxProfit;
        prevMaxProfit = currMaxProfit;
    }
    return prevMaxProfit;
}

File: kotlin/Dynamic Programming/NeighborhoodBurglaryOptimized.kt
fun neighborhoodBurglaryOptimized(houses: List<Int>): Int {
    if (houses.isEmpty()) {
        return 0
    }
    if (houses.size == 1) {
        return houses[0]
    }
    // Initialize the variables with the base cases.
    var prevMaxProfit = maxOf(houses[0], houses[1])
    var prevPrevMaxProfit = houses[0]
    for (i in 2 until houses.size) {
        val currMaxProfit = maxOf(prevMaxProfit, houses[i] + prevPrevMaxProfit)
        // Update the values for the next iteration.
        prevPrevMaxProfit = prevMaxProfit
        prevMaxProfit = currMaxProfit
    }
    return prevMaxProfit
}
